File: solutions/2026-03-02/solution.py
# Minimum Window Substring
# Difficulty: Hard
# URL: https://leetcode.com/problems/minimum-window-substring/

def minWindow(s: str, t: str) -> str:
  if not s or not t:
    return ""
  t_count = {}
  for char in t:
    t_count[char] = t_count.get(char, 0) + 1
  required_chars = len(t_count)
  left = 0
  min_len = float("inf")
  min_window = ""
  formed_chars = 0
  window_counts = {}
  for right in range(len(s)):
    if s[right] in t_count:
      window_counts[s[right]] = window_counts.get(s[right], 0) + 1
      if window_counts[s[right]] == t_count[s[right]]:
        formed_chars += 1
    while left <= right and formed_chars == required_chars:
      if right - left + 1 < min_len:
        min_len = right - left + 1
        min_window = s[left:right + 1]
      if s[left] in t_count:
        window_counts[s[left]] -= 1
        if window_counts[s[left]] < t_count[s[left]]:
          formed_chars -= 1
      left += 1
  return min_window

File: solutions/2026-03-02/solution.java
// Minimum Window Substring
// Difficulty: Hard
// URL: https://leetcode.com/problems/minimum-window-substring/

public String minWindow(String s, String t) {
  if (s == null || t == null || s.length() == 0 || t.length() == 0) {
    return "";
  }
  int[] t_count = new int[128];
  for (char c : t.toCharArray()) {
    t_count[c]++;
  }
  int required_chars = countNonZero(t_count);
  int left = 0;
  int min_len = Integer.MAX_VALUE;
  String min_window = "";
  int formed_chars = 0;
  int[] window_counts = new int[128];
  for (int right = 0; right < s.length(); right++) {
    if (t_count[s.charAt(right)] > 0) {
      window_counts[s.charAt(right)]++;
      if (window_counts[s.charAt(right)] == t_count[s.charAt(right)]) {
        formed_chars++;
      }
    }
    while (left <= right && formed_chars == required_chars) {
      if (right - left + 1 < min_len) {
        min_len = right - left + 1;
        min_window = s.substring(left, right + 1);
      }
      if (t_count[s.charAt(left)] > 0) {
        window_counts[s.charAt(left)]--;
        if (window_counts[s.charAt(left)] < t_count[s.charAt(left)]) {
          formed_chars--;
        }
      }
      left++;
    }
  }
  return min_window;
}
private int countNonZero(int[] arr) {
  int count = 0;
  for (int i : arr) {
    if (i > 0) {
      count++;
    }
  }
  return count;
}